### PR TITLE
State migration

### PR DESCRIPTION
**Pull request by pq**
Linter migration to a state representation (see #3896). (Blocked until a new analyzer is published.)

### CODE REVIEW

File: tool/doc.dart [hunk 1]
```diff
@@ -95,8 +96,7 @@ late Map<String, SinceInfo> sinceInfo;
 
 final Map<String, String> _fixStatusMap = <String, String>{};
 
-String describeMaturity(LintRule r) =>
-    r.maturity == Maturity.stable ? '' : ' (${r.maturity.name})';
+String describeMaturity(LintRule r) => r.state.isStable ? '' : r.state.label;
```
Comment (bwilkerson) [inline tool/doc.dart:99]: Consider renaming this to `describeState`.

File: tool/machine.dart [hunk 2]
```diff
@@ -43,7 +43,9 @@ String getMachineListing(Iterable<LintRule> ruleRegistry,
         'name': rule.name,
         'description': rule.description,
         'group': rule.group.name,
-        'maturity': rule.maturity.name,
+        // todo(pq): here for backwards compatibility -- should be removed.
```
Comment (bwilkerson) [inline tool/machine.dart:46]: What needs to happen to allow this to be removed?

**Review by bwilkerson — APPROVED** [review]
Are there any references to `maturity` that haven't been cleaned up, or does this allow us to remove it from the analyzer side?

File: tool/doc.dart (hunk 1 above)
Comment (pq) [reply tool/doc.dart:99]: Done. Thanks!

**Comment by pq** [comment]
> Are there any references to maturity that haven't been cleaned up, or does this allow us to remove it from the analyzer side?

Linter is clean w/ this PR.  After internal lints are updated we can remove it from analyzer.

**Pull request merged**